src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -483,6 +486,12 @@ def do_compile(self, invalidation_check, compile_contexts, classpath_product):
     invalid_targets = [vt.target for vt in invalidation_check.invalid_vts]
     valid_targets = [vt.target for vt in invalidation_check.all_vts if vt.valid]
 
+    if self.context.products.is_required_data("export_dep_as_jar_classpath"):
+      # filter modulized targets from invalid targets list.
+      target_root_addresses = [t.address for t in set(self.context.target_roots)]
+      dependees_of_target_roots = self.context.build_graph.transitive_dependees_of_addresses(target_root_addresses)
```

Comment:
Why are we baking this in rather than having the caller of the goal compute this? Doing this this way will "break" the contents of the `runtime_classpath`, because we won't have compiled everything that we were supposed to to compute that product.

Reply:
I see. It probably makes more sense to change `relevant_targets` to `non_dependees_of_target_roots` here conditionally: 
https://github.com/pantsbuild/pants/blob/9dea18f41dd5d1608375ddf0bc78742c04abcf35/src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py#L441-L450

and the condition should probably also check whether `runtime_classpath` is requested. In case both `runtime_classpath` and `export_dep_as_jar_classpath` are requested, then it should compile the whole thing.

Reply:
I tried specifying a compile goal and an export-dep-as-jar goal on the same invocation and nothing broke. Presumably because the compile targets were just used for export-dep-as-jar. The only way I could see this breaking would be if you can specify multiple goals with different target roots on one invocation. Is that possible?

Reply:
`export-dep-as-jar` and `test` together would probably break, because test will fail to find the classpath for `dependees_of_target_roots`.

Reply:
Okay, good to know. Just curious :)

Reply:
>Doing this this way will "break" the contents of the runtime_classpath, because we won't have >compiled everything that we were supposed to to compute that product

Maybe the solution is to just use the export_dep_as_jar_classpath as an inter task signal, and change the contents of runtime_classpath based on whether one or both of them are requested. Does this abuse the product_types mechanism?